math/core/sources/macros.move

```diff
@@ -180,41 +217,50 @@ public(package) macro fun clz<$Int>($value: $Int, $bit_width: u16): u16 {
     res
 }
 
-/// Multiply `a` and `b`, shift the product right by `shift`, and round according to `rounding_mode`.
+/// Compute the log in base 2 of a positive value with configurable rounding.
 ///
-/// This macro mirrors the ergonomics of `mul_div`, promoting the operands to `u256` and delegating to
-/// a shared helper that performs the computation using the most efficient implementation available.
-/// It starts from the floor of `(a * b) / 2^shift`, then applies the requested rounding mode. The
-/// overflow flag reports when the rounded value no longer fits in the `u256` range (i.e. significant
-/// bits remain above the lowest 256 bits after the shift).
+/// The algorithm first computes floor(log2(value)) using count-leading-zeros, then applies the
+/// requested rounding mode. Powers of 2 return exact results without additional rounding.
 ///
 /// #### Generics
 /// - `$Int`: Any unsigned integer type (`u8`, `u16`, `u32`, `u64`, `u128`, or `u256`).
 ///
 /// #### Parameters
-/// - `$a`, `$b`: Unsigned factors.
-/// - `$shift`: Number of bits to shift to the right. Must be less than 256.
+/// - `$value`: The unsigned integer to compute the logarithm for.
+/// - `$bit_width`: The bit width of the type (8, 16, 32, 64, 128, or 256).
 /// - `$rounding_mode`: Rounding strategy drawn from `rounding::RoundingMode`.
 ///
 /// #### Returns
-/// `(overflow, result)` where `overflow` reports that the rounded value cannot fit in 256 bits and
-/// `result` contains the rounded quotient when no overflow occurs.
-///
-/// #### Aborts
-/// Does not emit custom errors, but will inherit the Move abort that occurs when `$shift` is 256 or
-/// greater.
-public(package) macro fun mul_shr<$Int>(
-    $a: $Int,
-    $b: $Int,
-    $shift: u8,
+/// The base-2 logarithm as a `u16`, rounded according to the specified mode.
+/// Returns `0` if `$value` is 0.
+public(package) macro fun log2<$Int>(
+    $value: $Int,
+    $bit_width: u16,
     $rounding_mode: RoundingMode,
-): (bool, u256) {
-    let a_u256 = ($a as u256);
-    let b_u256 = ($b as u256);
-    let shift = $shift;
-    let rounding_mode = $rounding_mode;
-
-    mul_shr_inner(a_u256, b_u256, shift, rounding_mode)
+): u16 {
+    let (value, bit_width, rounding_mode) = ($value, $bit_width, $rounding_mode);
+    if (value == 0 as $Int) {
+        return 0
+    };
+    let zeros = clz!(value, bit_width);
+    let floor_log = bit_width - 1 - zeros;
+    if (rounding_mode == rounding::down()) {
+        return floor_log
+    };
+    let power_of_two = (1 as $Int) << (floor_log as u8);
+    if (value == power_of_two) {
+        return floor_log
+    };
+    if (rounding_mode == rounding::up()) {
+        return floor_log + 1
+    };
+    // Nearest: decide whether to round up or down based on midpoint (2^n × √2)
+    let threshold_exp = 2 * floor_log + 1;
+    if (log_should_round_up(value as u256, threshold_exp)) {
```

Comment:
The `let threshold_exp = 2 * floor_log + 1;` should be contained in the log_should_round_up function, we can pass `floor_log` as the second argument instead of `threshold_exp`.

Reply:
Then for the sake of clarity, let's create an inner function then, but still move the param internally, it is a matter of how easy is to understand why this works in 6 months.